File: interpolation_search.py
# implementation of interpolation search
# This is an enhance version of binary search
# Myambo Hlulani
# 20 May 2025
# Time complexity -> log(log(n))

def interpolation_search(array: list[int], target: int, left_pointer: int, right_pointer: int) -> int:
    # calculate for middle index
    middle_pointer = left_pointer + ((right_pointer - left_pointer) * (target - array[left_pointer]))// (array[right_pointer] - array[left_pointer])

    if left_pointer > right_pointer or array[left_pointer] == array[right_pointer] or middle_pointer > right_pointer:
        return -1
    if array[middle_pointer] == target:
        return middle_pointer
    if array[middle_pointer] < target:
        return interpolation_search(array, target, middle_pointer + 1, right_pointer)
    if array[middle_pointer] > target:
        return interpolation_search(array, target, left_pointer, middle_pointer - 1)

test_case: list[int] = [10, 13, 15, 16, 18, 19, 27, 28, 30, 32, 34, 39, 43, 45, 50]
LEFT = 0
RIGHT: int = len(test_case) - 1
TARGET: int = 10

print(interpolation_search(test_case, TARGET, LEFT, RIGHT)) # output = 0
print(interpolation_search(test_case, 0, LEFT, RIGHT)) # output = -1
print(interpolation_search(test_case, 30, LEFT, RIGHT)) # output = 8
print(interpolation_search(test_case, 50, LEFT, RIGHT)) # output = 14
print(interpolation_search(test_case, 27, LEFT, RIGHT)) # output = 6
print(interpolation_search(test_case, 200, LEFT, RIGHT)) # output = -1

File: InterpolationSearch.java
// implementation of Interpolation Search
// Myambo Hlulani
// 21 May 2025
// time-complexity -> O(log(log n))

public class InterpolationSearch {

    public static int search(int[] array, int target, int left_pointer, int right_pointer) {
        // if it is out of bound
        if (left_pointer > right_pointer || target > array[right_pointer] || target < array[left_pointer]) {
            return -1;
        }

        // handling ZeroDivisionError
        if (array[left_pointer] == array[right_pointer]) {
            return array[left_pointer] == target ? left_pointer : -1;
        }

        int middle = left_pointer + ((right_pointer - left_pointer) * (target - array[left_pointer]))
                / (array[right_pointer] - array[left_pointer]);

        // checking if the middle value is equal to the target
        if (array[middle] == target) {
            return middle;
        }

        // if target is less than middle value then shift right pointer into middle
        // index sub 1
        if (array[middle] > target) {
            return search(array, target, left_pointer, middle - 1);
        }

        // shifting the left pointer to middle index + 1 since the target is greater
        // than middle value
        return search(array, target, middle + 1, right_pointer);
    }

    public static void main(String[] args) {
        // array to be searched
        int[] array = { 10, 20, 30, 40, 50, 60, 500, 1000, 2000, 10000, 1111111 };

        // test cases
        System.out.println(search(array, 20, 0, array.length - 1)); // output = 1
        System.out.println(search(array, 0, 0, array.length - 1)); // output = -1
        System.out.println(search(array, 10, 0, array.length - 1)); // output = 0
        System.out.println(search(array, 1111111, 0, array.length - 1)); // output = 10
        System.out.println(search(array, 500, 0, array.length - 1)); // output = 6
    }
}
